src/tools/runk/libcontainer/src/cgroup.rs

```diff
@@ -3,38 +3,46 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
-use anyhow::{anyhow, Result};
-use rustjail::cgroups::fs::Manager as CgroupManager;
-use std::{
-    path::Path,
-    {fs, thread, time},
-};
+#[cfg(not(test))]
+use anyhow::anyhow;
+use anyhow::Result;
+use cgroups;
+use cgroups::freezer::{FreezerController, FreezerState};
+use std::{thread, time};
 
-pub fn destroy_cgroup(cgroup_mg: &CgroupManager) -> Result<()> {
-    for path in cgroup_mg.paths.values() {
-        remove_cgroup_dir(Path::new(path))?;
-    }
-
-    Ok(())
-}
-
-// Try to remove the provided cgroups path five times with increasing delay between tries.
-// If after all there are not removed cgroups, an appropriate error will be returned.
-fn remove_cgroup_dir(path: &Path) -> Result<()> {
-    let mut retries = 5;
-    let mut delay = time::Duration::from_millis(10);
+// Make sure we get a stable freezer state, so retry if the cgroup is still undergoing freezing.
+pub fn get_freezer_state(freezer: &FreezerController) -> Result<FreezerState> {
+    let mut retries = 10;
     while retries != 0 {
-        if retries != 5 {
-            delay *= 2;
-            thread::sleep(delay);
-        }
-
-        if !path.exists() || fs::remove_dir(path).is_ok() {
-            return Ok(());
+        let state = freezer.state()?;
+        match state {
+            FreezerState::Thawed => return Ok(FreezerState::Thawed),
+            FreezerState::Frozen => return Ok(FreezerState::Frozen),
+            FreezerState::Freezing => {
+                // sleep for 10 ms, wait for the cgroup to finish freezing
+                thread::sleep(time::Duration::from_millis(10));
+                retries -= 1;
+            }
         }
+    }
+    Ok(FreezerState::Freezing)
+}
 
-        retries -= 1;
+// check whether freezer state is frozon
+#[cfg(not(test))]
+pub fn is_paused(cgroup: &cgroups::Cgroup) -> Result<bool> {
+    let freezer_controller: &FreezerController = cgroup
+        .controller_of()
+        .ok_or_else(|| anyhow!("failed to get freezer controller"))?;
+    let freezer_state = get_freezer_state(freezer_controller)?;
+    match freezer_state {
+        FreezerState::Frozen => Ok(true),
+        _ => Ok(false),
     }
+}
 
-    return Err(anyhow!("failed to remove cgroups paths: {:?}", path));
+// non-root user has no permission to create cgroups in unit tests, so mock it and always return false
+#[cfg(test)]
+pub fn is_paused(_: &cgroups::Cgroup) -> Result<bool> {
+    Ok(false)
```

Comment:
Where is the creation operation? Do you mean `test_get_current_container_state`? 
Try to use a way like what agent has done:

- https://github.com/kata-containers/kata-containers/blob/f30fe86dc1f53b5e964e6f54de28a0ce0a818786/docs/Unit-Test-Advice.md#run-rust-tests-as-a-different-user
- https://github.com/kata-containers/kata-containers/blob/db2a4d6cdfe2d441be1260ea272c34a6f2e60fa9/src/agent/src/mount.rs#L1913

Reply:
Thanks for the reminder. I haven't noticed that the unit tests are run twice, once by the root user.